cloud/pkg/upgradecontroller/controller/downstream.go

```diff
@@ -0,0 +1,356 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachineryType "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	beehiveContext "github.com/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/cloud/pkg/common/client"
+	"github.com/kubeedge/kubeedge/cloud/pkg/common/messagelayer"
+	"github.com/kubeedge/kubeedge/cloud/pkg/common/modules"
+	"github.com/kubeedge/kubeedge/cloud/pkg/upgradecontroller/manager"
+	"github.com/kubeedge/kubeedge/common/constants"
+	commontypes "github.com/kubeedge/kubeedge/common/types"
+	"github.com/kubeedge/kubeedge/pkg/apis/upgrade/v1alpha1"
+	crdClientset "github.com/kubeedge/kubeedge/pkg/client/clientset/versioned"
+	crdinformers "github.com/kubeedge/kubeedge/pkg/client/informers/externalversions"
+)
+
+type DownstreamController struct {
+	kubeClient   kubernetes.Interface
+	crdClient    crdClientset.Interface
+	messageLayer messagelayer.MessageLayer
+
+	upgradeManager *manager.UpgradeManager
+}
+
+// Start DownstreamController
+func (dc *DownstreamController) Start() error {
+	klog.Info("Start Upgrade Downstream Controller")
+
+	go dc.syncUpgrade()
+
+	return nil
+}
+
+// syncUpgrade is used to get events from informer
+func (dc *DownstreamController) syncUpgrade() {
+	for {
+		select {
+		case <-beehiveContext.Done():
+			klog.Info("stop sync Upgrade")
+			return
+		case e := <-dc.upgradeManager.Events():
+			upgrade, ok := e.Object.(*v1alpha1.Upgrade)
+			if !ok {
+				klog.Warningf("object type: %T unsupported", e.Object)
+				continue
+			}
+			switch e.Type {
+			case watch.Added:
+				dc.upgradeAdded(upgrade)
+			case watch.Deleted:
+				dc.upgradeDeleted(upgrade)
+			case watch.Modified:
+				dc.upgradeUpdated(upgrade)
+			default:
+				klog.Warningf("upgrade event type: %s unsupported", e.Type)
+			}
+		}
+	}
+}
+
+func buildUpgradeResource(upgradeID, nodeID string) string {
+	resource := fmt.Sprintf("%s%s%s%s%s%s%s", "upgrade", constants.ResourceSep, upgradeID, constants.ResourceSep, "node", constants.ResourceSep, nodeID)
+	return resource
+}
+
+// upgradeAdded is used to process addition of new Upgrade in apiserver
+func (dc *DownstreamController) upgradeAdded(upgrade *v1alpha1.Upgrade) {
+	klog.V(4).Infof("add upgrade: %v", upgrade)
+	// store in cache map
+	dc.upgradeManager.UpgradeMap.Store(upgrade.Name, upgrade)
+
+	// If all or partial edge nodes upgrade is upgrading or completed, we don't need to send upgrade message
+	if isCompleted(upgrade) {
+		klog.Errorf("some upgrade is already upgrading or completed, don't send upgrade message again")
+		return
+	}
+
+	// get node list that need upgrading
+	var nodeName []string
+	if len(upgrade.Spec.NodeNames) != 0 {
```

Comment:
Could we set nodeNames and selector same time

Reply:
Now the nodeNames has a higher priority. So if we set both, we'll only take NodeNames into consideration.
I think we can add one restriction in admision so that we can only specify only one, but not two both